#37652856-Backend api implementation of project CRUD

tanerochris opened the pull request:
### Description
Developing the API backend for creating, updating and viewing of projects on voxnostra,
 ### How Has This Been Tested?
Tests have been written for the different endpoints
### Any background context you want to provide?
 ### Relevant Task link (from the project board) 
Card [37652856](https://github.com/orgs/BuildForSDG/projects/41#card-37652856)
### Questions
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have added necessary inline code documentation
- [x] I have added tests that prove my fix is effective and that this feature works
- [x] New and existing unit tests pass locally with my changes
closes #14

mernxl committed:
feat: add api error classes, implement catch all handlers in middlewares

mernxl committed:
feat(middlewares): add `json` method to `res` object

mernxl committed:
chore(deps): add `bcrypt`, `jsonwebtoken`, and `moment` packages

mernxl committed:
chore(props): add Authentication dependent properties to voxnostra.properties file

mernxl committed:
feat(libs): add authentication helpers, configure constants

mernxl committed:
fix(libs): fix db initialisation logic

mernxl committed:
fix(libs): fix session setup logic

mernxl committed:
refactor(middleware): add console.error logging for caught error

mernxl committed:
feat(libs): add `setUserSession` function to auth-helpers, fix config constant init

mernxl committed:
feat(schema): add password embedded schema, move user.model into user folder

mernxl committed:
feat(api): add user-authentication api endpoints

mernxl committed:
7-feat(api): swap testapi endpoint with a health-check endpoint

tanerochris committed:
6-docs(): Api design documentation
OpenAPI 3.0 design
Yaml and JSON files

closes #6

mernxl committed:
7-feat(schema): add email validation regex with onFail message

mernxl committed:
7-fix(middleware): fix sending 404 after response sent

- set content type for catchall error response

mernxl committed:
7-fix(api): fix res.session is undefined for reset-password

- workaround for JS Object destructuring inside an if block considered error by codacity

tanerochris committed:
6-docs(enpoints): updated API docs readme
Added instructions to run API endpoint documentation
Added link to external docs

close

tanerochris committed:
Merge branch 'ft/7-user-authentication' into ft/14-backend-project-crud

tanerochris committed:
14-feat(backend): Project CRUD API implementation

- Updated readme file
- Added schemas and model related to project feature
- Created an all model file to import into database ,prevents previous error from mongoose "cannot overrite model already compiled"
- Added ApiErrorResponse class to manage api errors
- Wrote tests for Project feature endpoints
- Updated setupTests.js with code to run before and after test suits run
- rename auth directory to user to match docs api endpoints

close #14

tanerochris committed:
14-feat(backend): Project CRUD API implementation

- Updated readme file
- Added schemas and model related to project feature
- Created an all model file to import into database ,prevents previous error from mongoose "cannot overrite model already compiled"
- Added ApiErrorResponse class to manage api errors
- Wrote tests for Project feature endpoints
- Updated setupTests.js with code to run before and after test suits run
- rename auth directory to user to match docs api endpoints

close #14

tanerochris committed:
14-feat(backend): Project CRUD API implementation

- Updated readme file
- Added schemas and model related to project feature
- Created an all model file to import into database ,prevents previous error from mongoose "cannot overrite model already compiled"
- Added ApiErrorResponse class to manage api errors
- Wrote tests for Project feature endpoints
- Updated setupTests.js with code to run before and after test suits run
- rename auth directory to user to match docs api endpoints

close #14

tanerochris committed:
14-feat(backend): Project CRUD API implementation

- added jest options --forceExit --detectOpenHandles --maxWorkers=10 to exit jest when all test pass

close #14

tanerochris committed:
14-fix(backend): Project CRUD API implementation

- fix codacy fail error
close #14

tanerochris committed:
14-fix(backend): Project CRUD API implementation

- fix codacy fail error
close #14

cliffordten commented:
Hey @tanerochris Please could you attach screenshot for test on different endpoints?

tanerochris commented:
Hello you can clone the branch and run the test. yarn test or npm test

On Fri, May 15, 2020, 19:56 Teneng Clifford <notifications@github.com>
wrote:

> Hey @tanerochris <https://github.com/tanerochris> Please could you attach
> screenshot for test on different endpoints?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BuildForSDG/voxnostra/pull/20#issuecomment-629424222>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACC3JXKVWUTHWCWLQDN34ELRRWF7LANCNFSM4NBFPPPQ>
> .
>

tanerochris committed:
14-refactor(structure): Rename folders and moved files

- documented project api endpoints
- refactored project structure, renamed folders and moved files raised in #16
- added prettier configuration

partially resolve #14 reviews

tanerochris committed:
14-refactor(structure): Rename folders and moved files

- documented project api endpoints
- refactored project structure, renamed folders and moved files raised in #16
- added prettier configuration

partially resolve #14 reviews

tanerochris committed:
14-refactor(structure): Rename folders and moved files

- documented project api endpoints
- refactored project structure, renamed folders and moved files raised in #16
- added prettier configuration

partially resolve #14 reviews

KaiserPhemi commented on pages/api/user/sign-out.js:
```diff
@@ -0,0 +1,19 @@
+ * @returns {Promise<void>}
+ */
+async function SignOutHandler({ req, res }) {
+  if (req.method === 'DELETE') {
```
# Nitpick
If this route is meant to specifically handle sign-out (cos the frontend should specifically send `delete` request in the first place) why check the http method in the first place? Besides I believe all checking/validation should be handle by a middleware.

tanerochris replied:
@mernxl  could you look into this.

KaiserPhemi commented on pages/api/user/sign-up.js:
```diff
@@ -0,0 +1,37 @@
+ * @returns {Promise<void>}
+ */
+async function SignUpHandler({ req, res }) {
+  if (req.method === 'POST') {
```
# Nitpick 
Same thing as the comment I left above.

KaiserPhemi commented on pages/api/projects/index.js:
```diff
@@ -0,0 +1,40 @@
+ */
+const IndexProjectHandler = async ({ req, res }) => {
+  if (req.method === 'GET') {
+    const limit = req.query.limit || 0;
```
# Nitpick
A better approach would be to abstract these into a helper file or a middleware

KaiserPhemi reviewed (changes requested):
# Critical

- File Naming: Going forward let's adopt `camelCase` for naming files as opposed to `snake-case` that's been used.
- It appears responses aren't sent for error messages thrown.
- I noticed we're using `res.json()` & having to set the header. `res.send()` automatically sets the header so it's preferable.
- I might need to schedule a sync with all today(21st May) on this .

tanerochris commented:
@KaiserPhemi I will love to mention that we are using NextJS for the
project ,it is an opinionated framework so there are some aspects of the
file structure and naming convention that is dictated by the framework.
Equally handling API requests.

On Thu, May 21, 2020, 03:46 Oluwafemi Akinwa <notifications@github.com>
wrote:

> *@KaiserPhemi* requested changes on this pull request.
> Critical
>
>    - File Naming: Going forward let's adopt camelCase for naming files as
>    opposed to snake-case that's been used.
>    - It appears responses aren't sent for error messages thrown.
>    - I noticed we're using res.json() & having to set the header.
>    res.send() automatically sets the header so it's preferable.
>    - I might need to schedule a sync with all today(21st May) on this .
>
> ------------------------------
>
> In pages/api/user/sign-out.js
> <https://github.com/BuildForSDG/voxnostra/pull/20#discussion_r428397670>:
>
> > @@ -0,0 +1,19 @@
> +import Middleware from '../../../middlewares';
> +/**
> + * Sets the session value to null.
> + *
> + * DELETE endpoint
> + * @param req
> + * @param res
> + * @returns {Promise<void>}
> + */
> +async function SignOutHandler({ req, res }) {
> +  if (req.method === 'DELETE') {
>
> Nitpick
>
> If this route is meant to specifically handle sign-out (cos the frontend
> should specifically send delete request in the first place) why check the
> http method in the first place? Besides I believe all checking/validation
> should be handle by a middleware.
> ------------------------------
>
> In pages/api/user/sign-up.js
> <https://github.com/BuildForSDG/voxnostra/pull/20#discussion_r428398945>:
>
> > +import { setUserSession } from '../../../helpers/auth-helpers';
> +import Middleware from '../../../middlewares';
> +import User from '../../../models/user/user.model';
> +
> +/**
> + * POST only
> + * Sign up a user, body must contain these two fields
> + *
> + * @param req
> + * @param req.body.email
> + * @param req.body.password
> + * @param res
> + * @returns {Promise<void>}
> + */
> +async function SignUpHandler({ req, res }) {
> +  if (req.method === 'POST') {
>
> Nitpick
>
> Same thing as the comment I left above.
> ------------------------------
>
> In pages/api/projects/index.js
> <https://github.com/BuildForSDG/voxnostra/pull/20#discussion_r428408301>:
>
> > @@ -0,0 +1,40 @@
> +import mongoose from 'mongoose';
> +import Middleware from '../../../middlewares';
> +
> +const Project = mongoose.model('Project');
> +/**
> + * @async
> + * @desc Api handles project list and search request
> + * @param {{ req: NextApiRequest, res: NextApiResponse }} prop
> + */
> +const IndexProjectHandler = async ({ req, res }) => {
> +  if (req.method === 'GET') {
> +    const limit = req.query.limit || 0;
>
> Nitpick
>
> A better approach would be to abstract these into a helper file or a
> middleware
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BuildForSDG/voxnostra/pull/20#pullrequestreview-415822168>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACC3JXNQTJJ5TXQGOWR4VLLRSSIY5ANCNFSM4NBFPPPQ>
> .
>
